### PR TITLE
Wait until race condition fix

### PR DESCRIPTION
**Pull request by AnthonyMDev**
This fixes issue #509. There was a race condition that resulted in the main RunLoop locking up, halting all tests indefinitely.

This occurred sometimes when `done()` was called asynchronously on a background thread.
We are checking if the result for the test is complete (ie. `done()` has been called) on line 258, then running the `RunLoop`. If the background thread marks the test complete between the check on line 258 and the running of the `RunLoop` on line 260, it stops the `RunLoop` **prior** to the `RunLoop` being started instead of after. This means the `RunLoop` is never stopped and the test hangs forever. This ensures that the completion of the test and the stopping of the `RunLoop` is called serially on the main thread, which prevents the race condition.

Unit test is included to verify the fix.

Checklist - While not every PR needs it, new features should consider this list:

 - [x] Does this have tests?
 - [x] Does this have documentation?
 - [ ] Does this break the public API (Requires major version bump)?
 - [ ] Is this a new feature (Requires minor version bump)?

### CODE REVIEW

File: Tests/NimbleTests/AsynchronousTest.swift [hunk 1]
```diff
@@ -165,6 +165,39 @@ final class AsyncTest: XCTestCase {
...
+        let timer = DispatchSource.makeTimerSource(flags: .strict, queue: timeoutQueue)
+        timer.schedule(
+            deadline: DispatchTime.now() + 5,
+            repeating: .never,
+            leeway: DispatchTimeInterval.milliseconds(1)
+        )
+        timer.setEventHandler {
+            failed = true
+            fail("Timed out: Main RunLoop stalled.")
+            CFRunLoopStop(CFRunLoopGetMain())
```
Comment (ikesyo) [inline Tests/NimbleTests/AsynchronousTest.swift]: `import CoreFoundation` is required to make it compilable on Linux.

Comment (AnthonyMDev) [reply Tests/NimbleTests/AsynchronousTest.swift]: Oh! Missed this. Yes, I'll make the change later today when I get a moment!

Comment (AnthonyMDev) [reply Tests/NimbleTests/AsynchronousTest.swift]: @ikesyo This is done!

**Comment by dipikasrivastava16** [comment]
Hey Guys.. I am facing the "main loop unresponsive" issue and hence keen to know, when this PR is going to be merged.. Currently I have applied the fix locally to my machine but would be helpful if this PR goes in to keep things dynamic.

**Comment by AnthonyMDev** [comment]
@ikesyo would you be able to approve and merge this now?

**Comment by paulz** [comment]
Could someone please make a minor release with this fix?
Our tests are hanging at random for a very long time which feels frustrating.

**Comment by ikesyo** [comment]
I'm sorry for the delay, [v8.0.6](https://github.com/Quick/Nimble/releases/tag/v8.0.6) has been released including the fix.

**Pull request merged**